File: src/models/rioCharacter.ts
import CharacterIdentity from './characteridentity';

const raidName = require('../data/raidData.json').lastRaid;

export default class RioCharacter {
  public Name: string;
  public Race: string;
  public Class: string;
  public Active_spec_name: string;
  public Active_spec_role: string;
  public Gender: string;
  public Faction: string;
  public Achievement_points: number;
  public Honorable_kills: number;
  public Thumbnail_url: string;
  public Region: string;
  public Realm: string;
  public Profile_url: string;
  public Profile_banner: string;
  public Gear_item_level_total: number;
  public Mythic_plus_score_all: number;
  public Mythic_plus_score_dps: number;
  public Mythic_plus_score_healer: number;
  public Mythic_plus_score_tank: number;
  public Raid_progression_summary: string;
  public Mythic_plus_weekly_highest_level_runs_level: number;
  public Mythic_plus_weekly_highest_level_runs_score: number;
  public Mythic_plus_weekly_highest_level_runs_name: string;
  public Mythic_plus_weekly_highest_level_runs_short_name: string;
  public Mythic_plus_weekly_highest_level_runs_upgrade: number;
  public Mythic_plus_weekly_highest_level_runs_time: number;
  public Mythic_plus_season_highest_level_runs_level: number;
  public Mythic_plus_season_highest_level_runs_score: number;
  public Mythic_plus_season_highest_level_runs_name: string;
  public Mythic_plus_season_highest_level_runs_short_name: string;
  public Mythic_plus_season_highest_level_runs_upgrade: number;
  public Mythic_plus_season_highest_level_runs_time: number;
  
  public constructor(data: any) {
    this.Name = data.name;
    this.Class = data.class;
    this.Thumbnail_url = data.thumbnail_url;
    this.Profile_url = data.profile_url;
    this.Gear_item_level_total = data.gear.item_level_total;
    const score = data.mythic_plus_scores_by_season[0].scores;
    this.Mythic_plus_score_all = score.all;
    this.Mythic_plus_score_dps = score.dps;
    this.Mythic_plus_score_healer = score.healer;
    this.Mythic_plus_score_tank = score.tank;
    this.Raid_progression_summary = data.raid_progression[raidName].summary;
    const week_keys: Array<any> = data.mythic_plus_weekly_highest_level_runs;
    if (week_keys.length > 0) {
      const best_week_key = week_keys.sort((a, b) => b.score - a.score)[0];
      this.Mythic_plus_weekly_highest_level_runs_level = best_week_key.mythic_level;
      this.Mythic_plus_weekly_highest_level_runs_score = best_week_key.score;
      this.Mythic_plus_weekly_highest_level_runs_name = best_week_key.dungeon;
      this.Mythic_plus_weekly_highest_level_runs_short_name = best_week_key.short_name;
      this.Mythic_plus_weekly_highest_level_runs_upgrade = best_week_key.num_keystone_upgrades;
      this.Mythic_plus_weekly_highest_level_runs_time = best_week_key.clear_time_ms;
    }
    const season_keys: Array<any> = data.mythic_plus_highest_level_runs;
    if (season_keys.length > 0) {
      const best_season_key = season_keys.sort((a, b) => b.score - a.score)[0];
      this.Mythic_plus_season_highest_level_runs_level = best_season_key.mythic_level;
      this.Mythic_plus_season_highest_level_runs_score = best_season_key.score;
      this.Mythic_plus_season_highest_level_runs_name = best_season_key.dungeon;
      this.Mythic_plus_season_highest_level_runs_short_name = best_season_key.short_name;
      this.Mythic_plus_season_highest_level_runs_upgrade = best_season_key.num_keystone_upgrades;
      this.Mythic_plus_season_highest_level_runs_time = best_season_key.clear_time_ms;
    }
  }

  public static getCharRaiderIoUrl(char: CharacterIdentity): string {
    const name = encodeURIComponent(char.Name);
    const fields = 'gear%2Craid_progression%2Cmythic_plus_scores_by_season%3Acurrent%2Cmythic_plus_weekly_highest_level_runs%2Cmythic_plus_highest_level_runs';
    return `https://raider.io/api/v1/characters/profile?region=${char.Region}&realm=${char.Realm}&name=${name}&fields=${fields}`;
  }
}

File: Models/RioCharacter.cs
namespace RuthlessBoreas.Models
{
  using RuthlessBoreas.Models.Dto;
  using RuthlessBoreas.Services;
  using System.Linq;

  public class RioCharacter
  {
    public string Class { get; }
    public string Portrait { get; }
    public string RioProfile { get; }
    public int ItemLevel { get; }
    public int ScoreAll { get; }
    public int ScoreDps { get; }
    public int? ScoreHealer { get; }
    public int? ScoreTank { get; }
    public string RaidProgress { get; }
    public string MainName { get; }
    public Dungeon MaxWeekKey { get; }
    public Dungeon MaxSeasonKey { get; }

    public RioCharacter(DtoRaiderIo dto)
    {
      this.Class = dto.Class;
      this.Portrait = dto.Thumbnail_url;
      this.RioProfile = dto.Profile_url;
      this.ItemLevel = dto.Gear.Item_level_total > 0 ? dto.Gear.Item_level_total : dto.Gear.Item_level_equipped;
      var score = dto.Mythic_plus_scores_by_season[0].Scores;
      this.ScoreAll = (int)score.All;
      this.ScoreDps = (int)score.Dps;
      this.ScoreHealer = Utils.CanHeal(this.Class) ? (int?)score.Healer : null;
      this.ScoreTank = Utils.CanTank(this.Class) ? (int?)score.Tank : null;
      this.RaidProgress = dto.Raid_progression[StorageService.LAST_RAID].Summary;

      if (dto.Mythic_plus_weekly_highest_level_runs.Length > 0)
      {
        var bestWeekKey = dto.Mythic_plus_weekly_highest_level_runs.OrderByDescending(run => run.Score).First();
        this.MaxWeekKey = new Dungeon(bestWeekKey.Dungeon, bestWeekKey.Short_name, (int)bestWeekKey.Score,
          bestWeekKey.Mythic_level, bestWeekKey.Num_keystone_upgrades, bestWeekKey.Clear_time_ms);
      }

      if (dto.Mythic_plus_highest_level_runs.Length > 0)
      {
        var bestSeasonKey = dto.Mythic_plus_highest_level_runs.OrderByDescending(run => run.Score).First();
        this.MaxSeasonKey = new Dungeon(bestSeasonKey.Dungeon, bestSeasonKey.Short_name, (int)bestSeasonKey.Score,
          bestSeasonKey.Mythic_level, bestSeasonKey.Num_keystone_upgrades, bestSeasonKey.Clear_time_ms);
      }
    }
  }
}
